### PR TITLE
Add GRPOTrainModule subclassing TransformerTrainModule (GRPO olmo-core: PR 2 of 5)

### PR DESCRIPTION
**Pull request by finbarrtimbers**
Adds a `GRPOTrainModule` that subclasses OLMo-core's `TransformerTrainModule`
- Inherits: `optim_step`, `zero_grads`, `eval_batch`, `state_dict`, `load_state_dict`, with minor changes to support the reference policy
- Overrides: `train_batch` with GRPO loss computation

We refactor some of the code in `grpo_fast.py` into `grpo_utils.py` to share it between the implementations. 

Runs: 

- Single GPU GRPO: [Beaker](https://beaker.org/orgs/ai2/workspaces/open-instruct-dev/work/01KFNZ3Y7S07SD0FYW9CGE4TPK?taskId=01KFNZ3Y80DW497DYKDG7QYXSX&jobId=01KFNZ3YBRSFAEEF90P0P4AVS7)
- Multi-node GRPO: [Beaker](https://beaker.org/orgs/ai2/workspaces/open-instruct-dev/work/01KFP1NWA4P3KDBX6V3J8HGRQ9?taskId=01KFP1NWADYH8S8RJDV5ADGYXG&jobId=01KFP1NX1X9QC1XQ9FVT59KHG4)

GPU_TESTS=[01KFPGA7V0JDYT0HQH4N44RQT0](https://beaker.org/orgs/ai2/workspaces/open-instruct-dev/work/01KFPGA7V0JDYT0HQH4N44RQT0?taskId=01KFPGA7V58J2YFZ0SS44YY2NB&jobId=01KFPGA7YSWCQRYH55SAN4MKBT)

### CODE REVIEW

File: open_instruct/olmo_core_train_modules.py [hunk 1]
```diff
@@ -0,0 +1,193 @@
...
+                pg_losses, pg_losses2, pg_loss, kl = grpo_utils.compute_grpo_loss(
+                    new_logprobs=new_logprobs,
+                    ratio=ratio,
+                    advantages=advantages[:, 1:],
+                    ref_logprobs=ref_logprobs_BT[sample_idx] if ref_logprobs_BT is not None else None,
+                    config=self.grpo_config,
+                )
+                loss = pg_loss + self.grpo_config.beta * kl
+
+                num_tokens = response_mask.sum()
```
Comment (hamishivi) [inline open_instruct/olmo_core_train_modules.py]: I don't think this is right for num_tokens - it should be the total token count across the entire batch (including all grad accum steps and including all samples across the DP dimension)

File: open_instruct/olmo_core_train_modules.py [hunk 2]
```diff
@@ -0,0 +1,193 @@
...
+            else:
+                ref_logprobs_BT = None
+
+        with torch.no_grad():
+            old_logprobs_BT = grpo_utils.compute_logprobs(
+                self.model, data_BT, self.pad_token_id, self.grpo_config.temperature, use_grad=False
+            )
+
+        num_samples = len(data_BT.query_responses)
+        accumulation_steps = max(num_samples // self.grpo_config.num_mini_batches, 1)
```
Comment (hamishivi) [inline open_instruct/olmo_core_train_modules.py]: I think we want math.ceil here, right? E.g., If I set num_mini_batches to 2, and I get 13 samples, then I throw out almost half my batch. We normalise the loss by the number of tokens in the batch so things should scale appropriately?

Or we leave it as is, but we should comment in that we are doing `drop_last` style behaviour.

File: open_instruct/grpo_utils.py [hunk 3]
```diff
@@ -215,3 +229,93 @@ def __post_init__(self):
...
+    pad_token_id: int,
+    temperature: float,
+    use_grad: bool = False,
+) -> list[torch.Tensor]:
+    """Compute log probabilities for all samples in batch."""
+    logprobs_BT: list[torch.Tensor] = []
+
+    context = torch.enable_grad() if use_grad else torch.no_grad()
+    with context:
+        for i in range(len(data_BT.query_responses)):
```
Comment (hamishivi) [inline open_instruct/grpo_utils.py]: should we have a logprob bsz arg? Just a random thought.

File: open_instruct/olmo_core_train_modules.py [hunk 4]
```diff
@@ -0,0 +1,193 @@
...
+                        self.optim_step()
+                    self.zero_grads()
+
+        if local_step % accumulation_steps != 0:
+            if not dry_run:
+                self.optim_step()
+            self.zero_grads()
+
+        if not dry_run and num_steps > 0:
+            self.record_metric("train/pg_loss", (total_pg_loss / num_steps).item(), ReduceType.mean)
```
Comment (hamishivi) [inline open_instruct/olmo_core_train_modules.py]: Similarly here, I'm not sure how olmo-core does it, but to be correct we should weight each rank by the number of tokens they process, which can be variable across diff DP/SP/CP ranks.

File: open_instruct/olmo_core_train_modules.py (hunk 1 above)
Comment (finbarrtimbers) [reply open_instruct/olmo_core_train_modules.py]: Fixed!

File: open_instruct/olmo_core_train_modules.py (hunk 2 above)
Comment (finbarrtimbers) [reply open_instruct/olmo_core_train_modules.py]: Fixed!

File: open_instruct/grpo_utils.py (hunk 3 above)
Comment (finbarrtimbers) [reply open_instruct/grpo_utils.py]: Good idea! Done.

File: open_instruct/olmo_core_train_modules.py (hunk 4 above)
Comment (finbarrtimbers) [reply open_instruct/olmo_core_train_modules.py]: Fixed!

File: CHANGELOG.md [hunk 5]
```diff
@@ -4,6 +4,15 @@ All notable changes to this project will be documented in this file.
 
 
 ### Added
+- Added retry logic with exponential backoff to `make_api_request` for tool API calls (retries on timeouts, connection errors, 429, and 5xx). Also added configurable `max_concurrency` parameter to tool configs for controlling Ray actor concurrency per-tool. https://github.com/allenai/open-instruct/pull/1388
```
Comment (hamishivi) [inline CHANGELOG.md]: I think this changelog edit is wrong! Probably a wonky merge :D
